runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/ActiveWorkState.java

```diff
@@ -120,6 +123,42 @@ synchronized ActivateWorkResult activateWorkForKey(ShardedKey shardedKey, Work w
     return ActivateWorkResult.QUEUED;
   }
 
+  public static class FailedTokens {
+    public long workToken;
+    public long cacheToken;
+
+    public FailedTokens(long workToken, long cacheToken) {
+      this.workToken = workToken;
+      this.cacheToken = cacheToken;
+    }
+  }
+
+  synchronized void failWorkForKey(Map<Long, List<FailedTokens>> failedWork) {
+    // Note we can't construct a ShardedKey and look it up in activeWork directly since
+    // HeartbeatResponse doesn't include the user key.
+    for (Entry<ShardedKey, Deque<Work>> entry : activeWork.entrySet()) {
+      List<FailedTokens> failedTokens = failedWork.get(entry.getKey().shardingKey());
+      if (failedTokens == null) continue;
+      for (FailedTokens failedToken : failedTokens) {
+        for (Work queuedWork : entry.getValue()) {
+          WorkItem workItem = queuedWork.getWorkItem();
+          if (workItem.getWorkToken() == failedToken.workToken
+              && workItem.getCacheToken() == failedToken.cacheToken) {
+            LOG.error(
```

Comment:
warning/info? users don't like errors

We've also added reassuring text like, "This happens during autocaling and backend events, the data will be reprocessed and not lost" in other cases and might want somethign like that here too.

Reply:
I'll make it a debug message for now so it doesn't show up all the time.